File: snippets/ts/src/foundations/services/basic_service.ts
import * as restate from "@restatedev/restate-sdk";

type SubscriptionRequest = {
  userId: string;
  creditCard: string;
  subscriptions: string[];
};

function createRecurringPayment(creditCard: string, paymentId: string) {
  return undefined;
}

function createSubscription(userId: string, subscription: string, payRef: any) {
  return undefined;
}

// <start_here>
const subscriptionService = restate.service({
  name: "SubscriptionService",
  handlers: {
    add: async (ctx: restate.Context, req: SubscriptionRequest) => {
      const paymentId = ctx.rand.uuidv4();

      const payRef = await ctx.run(() =>
        createRecurringPayment(req.creditCard, paymentId)
      );

      for (const subscription of req.subscriptions) {
        await ctx.run(() =>
          createSubscription(req.userId, subscription, payRef)
        );
      }
    },
  },
});
// <end_here>

export default subscriptionService;

File: snippets/go/foundations/services/basic_service.go
package main

import (
	restate "github.com/restatedev/sdk-go"
)

type SubscriptionRequest struct {
	UserId        string   `json:"userId"`
	CreditCard    string   `json:"creditCard"`
	Subscriptions []string `json:"subscriptions"`
}

func createRecurringPayment(creditCard string, paymentId string) (string, error) {
	return "", nil
}

func createSubscription(userId string, subscription string, payRef string) (string, error) {
	return "", nil
}

// <start_here>
type SubscriptionService struct{}

func (SubscriptionService) Add(ctx restate.Context, req SubscriptionRequest) error {
	paymentId := restate.UUID(ctx).String()

	payRef, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
		return createRecurringPayment(req.CreditCard, paymentId)
	})
	if err != nil {
		return err
	}

	for _, subscription := range req.Subscriptions {
		_, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
			return createSubscription(req.UserId, subscription, payRef)
		})
		if err != nil {
			return err
		}
	}

	return nil
}

// <end_here>
